src/adapters/methods/bottleneck.py

```diff
@@ -0,0 +1,372 @@
+from typing import Dict, List, Mapping, NamedTuple, Optional, Union
+
+import torch
+from torch import nn
+
+from ..composition import (
+    AdapterCompositionBlock,
+    Average,
+    BatchSplit,
+    Fuse,
+    Parallel,
+    Split,
+    Stack,
+    adjust_tensors_for_parallel,
+)
+from ..configuration import BnConfig
+from ..context import ForwardContext
+from .adapter_layer_base import ComposableAdapterLayerBase
+from .modeling import Adapter, BertFusion, ParallelAdapter
+
+
+class BottleneckState(NamedTuple):
+    """
+    Models the input and output states of a bottleneck adapter layer.
+
+    Args:
+        hidden_states (torch.Tensor): The layer input/ output hidden states.
+        input_tensor (torch.Tensor): The Transformer sub-block residual connection inputs.
+        adapter_residual (torch.Tensor): The adapter residual connection inputs.
+        layer_norm (torch.nn.Module, optional): The Transformer layer norm module.
+        bottleneck_up (torch.Tensor, optional):
+            The up-projected bottleneck MLP output. This is only for Fuse compositions.
+    """
+
+    hidden_states: torch.Tensor
+    input_tensor: torch.Tensor
+    adapter_residual: torch.Tensor
+    layer_norm: Optional[torch.nn.Module]
+    bottleneck_up: Optional[torch.Tensor] = None
+
+
+class BottleneckLayer(ComposableAdapterLayerBase, nn.Module):
+    adapter_modules_name = "adapters"
```

Comment:
This name reflects the name of the module dict holding adapter weights. Changing this would break all existing adapter checkpoints, so I'd prefer to keep it as is.

Reply:
Ok then let's keep it as it is